### PR TITLE
fix: Update service polling

### PR DESCRIPTION
**Pull request by averikitsch**

### CODE REVIEW

**Comment by averikitsch** [comment]
@bharathkkb this PR and PR #30 are reliant on each other; however I felt it was easier for separate reviews.

File: src/cloudRun.ts [hunk 1]
```diff
@@ -299,4 +303,58 @@ export class CloudRun {
       this.methodOptions,
     );
   }
+
+  async pollService(serviceResponse: run_v1.Schema$Service): Promise<string> {
+    let url = getUrl(serviceResponse);
+    while (!getReadyStatus(serviceResponse) && !url) {
```
Comment (bharathkkb) [inline src/cloudRun.ts:309]: Shouldn't we have a timeout incase it is never ready?

File: src/cloudRun.ts [hunk 2]
```diff
@@ -299,4 +303,58 @@ export class CloudRun {
...
+}
+
+function getReadyStatus(serviceResponse: run_v1.Schema$Service): boolean {
+  const revisionName = get(serviceResponse, 'spec.template.metadata.name');
+  const createdRevision = get(
+    serviceResponse,
+    'status.latestCreatedRevisionName',
+  );
+  const latestRevision = get(serviceResponse, 'status.latestReadyRevisionName');
+  if (revisionName) {
```
Comment (bharathkkb) [inline src/cloudRun.ts:326]: Comments on the logic when it evaluates to true would be useful to understand what are the ready states.

File: src/cloudRun.ts (hunk 1 above)
Comment (averikitsch) [reply src/cloudRun.ts]: done

**Comment by averikitsch** [comment]
@bharathkkb can you re-review when you have time?

File: src/cloudRun.ts [hunk 3]
```diff
@@ -299,4 +303,74 @@ export class CloudRun {
...
+    'status.traffic',
+  );
+  let revision;
+  if (traffic) {
+    revision = traffic.find((revision) => {
+      return revision.revisionName == revisionName && revision.url;
+    });
+  }
+  // Or return service url
+  return get(revision, 'url') || get(serviceResponse, 'status.url') || '';
```
Comment (unrealcloud) [inline src/cloudRun.ts:375]:
```suggestion
  return get(revision, 'url') || get(serviceResponse, 'status.url', '');
```

File: src/cloudRun.ts [hunk 4]
```diff
@@ -299,4 +303,74 @@ export class CloudRun {
...
+  );
+  const latestRevision = get(serviceResponse, 'status.latestReadyRevisionName');
+  // Latest created revision must equal latest ready revision
+  if (revisionName) {
+    // After first deployment, revision name is set
+    return (
+      revisionName &&
+      createdRevision &&
+      latestRevision &&
+      revisionName == createdRevision &&
```
Comment (unrealcloud) [inline src/cloudRun.ts:345]:
```suggestion
      createdRevision == createdRevision &&
```

**Review by unrealcloud — APPROVED** [review]

**Review by bharathkkb — APPROVED** [review]
Apologies for the late review, got buried in emails :)

**Comment by Rahma-AlWadhahi** [comment]
Hello @averikitsch, @unrealcloud and @bharathkkb.

> @bharathkkb this PR and PR #30 are reliant on each other; however I felt it was easier for separate reviews.

 I have noticed that one of the two PRs was merged (#30) while the other is not yet merged (#31). I am facing the same issue which was mentioned in #36 . Could you please take a look at it?

**Pull request merged**